src/java.base/share/classes/java/security/Provider.java

```diff
@@ -1267,22 +1312,19 @@ public Service getService(String type, String algorithm) {
      *
      * @since 1.5
      */
-    public synchronized Set<Service> getServices() {
+    public Set<Service> getServices() {
         checkInitialized();
-        if (legacyChanged || servicesChanged) {
-            serviceSet = null;
-        }
-        if (serviceSet == null) {
-            ensureLegacyParsed();
```

Comment:
Hi @valeriepeng! I believe that with this change, `getServices()` will return invalid legacy services. Before we called `ensureLegacyParsed()`, which eventually called `removeInvalidServices()`. In `getService(String, String)`, we are now explicitly checking for `isValid()` to keep the old behavior. Shouldn't we do something similar here as well? Am I missing something or is this an intended change?

Reply:
Hmm, could be. Let me check into it and I will have to file a separate bug to address this since the changes have already been integrated. Thanks for the comments.

Reply:
Thanks for looking into this! Please let me if you open a new bug so we can track it.

Reply:
Filed https://bugs.openjdk.org/browse/JDK-8292739
A PR should be out in a day or two.